GX-TimeFrequency/src/minimax_tau.F90

```diff
@@ -2907,50 +2906,36 @@ subroutine get_points_weights_tau(grid_size, e_range, ac_we, ierr)
     integer, intent(out)                                   :: ierr
 
     !> Internal variables
-    integer                                                :: ien, kloc, bup
+    integer                                                :: kloc, bup
     type(er_aw_aux)                                        :: aw
     real(kind=dp)                                          :: e_ratio
 
-
     !> Begin work
     ierr = 0
+    if (.not. any(tau_npoints_supported == grid_size)) then
+       ierr = 1
+       _REGISTER_EXC("The grid size you chose is not available.")
+       return
+    end if
 
-    if (any(tau_npoints_supported == grid_size)) then
-       ! Find location of grid size
-       kloc = findloc(tau_npoints_supported, grid_size, 1)
-       bup = energy_ranges_grids(kloc)
-
-       ! Allocate and set type elements
-       allocate(aw%energy_range(bup))
-       allocate(aw%aw_erange_matrix(2*grid_size, bup+1))
-       call set_aw_array(grid_size, aw)
-
-
-       ! Select energy region with binary search
-       ien = find_erange(bup, aw%energy_range, e_range)
-
-       ! Scale grids for large sizes when erange falls in the
-       ! first energy range
-       e_ratio = 1.0_dp
-       if (ien == 1 .and. grid_size > 20) then
-          e_ratio = aw%energy_range(1)/e_range
-          if (e_ratio > 1.5_dp) then
-             e_ratio = e_ratio/1.5_dp
-          endif
-       end if
+    ! Find location of grid size
+    kloc = findloc(tau_npoints_supported, grid_size, 1)
+    bup = energy_ranges_grids(kloc)
 
-       ac_we(:) = aw%aw_erange_matrix(:, ien)
+    ! Allocate and set type elements
+    allocate(aw%energy_range(bup))
+    allocate(aw%aw_erange_matrix(2*grid_size, bup+1))
+    call set_aw_array_tau(grid_size, aw)
```

Comment:
These should all be included in a constructor for the class:

`call aw%init(tau_npoints_supported, grid_size)
`
which would look something like

```fortran
subroutine init(this, tau_npoints_supported, grid_size)
    ...
  this%grid_size = grid_size
  this%tau_npoints_supported = tau_npoints_supported
  kloc = findloc(tau_npoints_supported, grid_size, 1)
  bup = energy_ranges_grids(kloc)
  allocate(this%energy_range(bup))
  allocate(this%aw_erange_matrix(2*grid_size, bup+1))
  call this%set_aw_array_tau()
end subroutine
```

else one has some arbitrary mix and match of OO and functions

Reply:
Here I struggled to get a clean solution. The things is, we have set_aw_array_tau() and set_aw_array_omega(), and this are defined in two different modules that use the útiles module. This creates a cyclic dependence if we want to add this setters to the constructor. What do you think?

Reply:
keep in mind,structural changes now requires more testing and my fork was heavily tested by Maryam, Antonio and myself.